### ISSUE TITLE
"Failed to sync bookmarks" with a big buku  bookmarks.db (10k)

### ISSUE THREAD

**Issue by m040601** [SamHH/bukubrow-webext#116]
bukubrow 4.0.1 from AUR archlinux
bukubrow-webext from, https://addons.mozilla.org/en-US/firefox/addon/bukubrow/
firefox 67.0
ArchLinux   5.1.3-arch2-1-ARCH 
x86_64 GNU/Linux

I never managed to get bukubrow+bukubrow-webextension working with my buku bookmarks.db containing 10000 bookmarks with the file size of 4.6 M.
I can use it with 1000 bookmarks.

"bukubrow --install-firefox" installs everything correctly.
I tried waiting some days for a new version, tried the precompiled version, tried several times with fresh firefox profiles with default settings etc.

I dont have any performance bottleneck issues using a big buku database with the standard buku cli tools.
I have several other extensions making use of native messanging (ex. browserpass) , none of them give me any problems.

1. Is there any size limit being imposed by bukubrow/webext ?
2. Other 3rd party buku tools (ex buku_run, rofi , dmenu helpers, ) dont have this problem with big buku databases (what is considered big anyway ?) although they might work more slugish .
3. If I were to use different buku bookmark.db' s in different locations (by splitting a big one into small ones), would a future bukubrow support this kind of usage ?
4. Is there any logging or "debugging" option in firefox that I can use to provide you with some more specific details of this error message ?
5. I'm curious about other peoples' bukubrow/webext performance on firefox. Maybe you could ask the users for some detailed feedback on this. How big is your buku bookmarks.db ? Do you see Firefox struggilng with RAM/CPU ? At which point ? Searching ? Listing ? Adding ?What would be the theoretical limit of confortably  using this firefox/bukubrow combo as an external bookmarks manager ? Is the bottleneck in bukubrow or the webext ?

**Comment by SamHH** [comment]
Hey, I theorised a while ago that someone would eventually report this issue due to the 1MB message limit from the native messenger to the webext. There's the relevant issue [here](https://github.com/SamHH/bukubrow-host/issues/4) that I'll now prioritise. Thanks for the report!

**Comment by m040601** [comment]
aaarrrrrrghhh ..... I completely missed, that you already had this "large Buku databases" issues open. Sorry for that. Anyway I do know completely understand the develelopment challenge .

**Comment by SamHH** [comment]
Not sure when the release will be for this, but support has been added [here (webext)](https://github.com/SamHH/bukubrow-webext/commit/7b5d0b1701667e68c7b4ad5062ccf03cbb88981d) and [here (host)](https://github.com/SamHH/bukubrow-host/commit/3f2b5cd7de7eabb8c14ead74c9483f03b13ff138).

If you know how to compile each project and test locally that'd be super helpful, else, no matter how much testing I do against my 25kB database, it'll be a bit of a Hail Mary. If not no worries. :slightly_smiling_face:

**Comment by SamHH** [comment]
A release with this fix included is going out today, let me know if it works for you. :slightly_smiling_face:

**Comment by m040601** [comment]
great job !

i still need to do more thorough testing, but here goes
some quick feedback with version 5.0.0.2:

1. With 1000 bookmarks 370kb bookmarks.db database file
responds better than previous version, much smoother feeling , snappy

 2.  With 10000 bookmarks 4.7mega bookmarks.db database file
great succezzzz ! at least no more error message.
firefox struggles alot to load. CPU usage spikes during this time when bukubrow/firefox are "talking" with buku.
took a looooooong time to load, but I even managed to search for tags. And I could even add and tag new bookmark new .
 Seems it wrote everything nicely to the database. No corruption issues so far.

**Comment by SamHH** [comment]
The performance hit here could be coming from one or both of two places.

First, the host needs to serialise every single database entry it finds to JSON, and I think my initial implementation is naive and results in it serialising everything twice. On this point, the WebExt did used to cache the output from the host, though I removed that in favour of it implicitly syncing on load.

Second, the WebExt performs some sorting and such across the bookmarks whenever you change the filter. It memoizes what it can, but JavaScript is never going to perform particularly well looping through an array of such a big size.

One solution to this would be to possibly move all of the sorting, filtering, etc logic to the host, and instead query the host from the frontend. This would mean potentially less serialisation (particularly if we only serialise beyond ten bookmarks when the load more button has been clicked), as well as being able to query the database directly as opposed to fetching everything, serialising it, and then looping through and querying that vector/array.

I've made a ticket [here](https://github.com/SamHH/bukubrow-host/issues/16) to investigate the host serialisation as this would be a relatively easy win.
